### PR TITLE
Fix a ``MemoryError`` in ``sphinx.ext.intersphinx`` when using ``None`` or ``typing.*`` as inline type references.

### LINKED ISSUE #11337: [BUG] ``MemoryError`` in ``sphinx.ext.intersphinx`` and undesirable side-effects in domains.
### Describe the bug

As mentioned in https://github.com/sphinx-doc/sphinx/issues/11282#issuecomment-1511900129, I found that `sphinx.ext.intersphinx` has a serious bug when parsing a document with the following structure:

```rst

.. py:function:: foo(x)
   
   The foo function.
   
   :param x: Some x.
   :type x: typing.Any

.. py:function:: bar(a)
   
   The bar function.
   
   :param a: Some a.
   :type a: typing.Any
```

As I said, here what happens during the resolution phase:

https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L1328-L1331

In particular, when encountering `typing.*` or ``None`` in an inline reference such as `:type NAME: TYPE`, the latter will be considered as an `obj` reference (and not a `class` reference). This is due to

https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L96-L100

Now, if `sphinx.ext.intersphinx` is enabled, we have:

https://github.com/sphinx-doc/sphinx/blob/5a6b2b16b6b588cdd4dab1b43b7873153da10baa/sphinx/ext/intersphinx.py#L355-L370

I found the following issues:

- If the role is `py:obj`, the corresponding object types are all possible object types. In particular, `'attribute'` and `'method'` coexist and `'method'` will be duplicated. This can lead to a ``MemoryError`` (I had a medium-size project where `object_types` grow more than of a hundred times so I can imagine when there are VERY large projects).
- In https://github.com/sphinx-doc/sphinx/blob/5a6b2b16b6b588cdd4dab1b43b7873153da10baa/sphinx/ext/intersphinx.py#L422-L429 the `objtypes` list is `domain._role2type[typ]`. This means that any modification to the former propagate to the latter, thereby causing an undesirable side-effect. Note that this **currently** seems to only affect the `py:obj` role *but* the side-effect can also be seen for other property-like roles.

With my above example, the domain's role dispatcher map *grows* as follows:

```python
# domain._role2type['obj'] when parsing foo()
['function', 'data', 'class', 'exception', 'method', 'classmethod', 'staticmethod', 'attribute', 'property', 'module']

# domain._role2type['obj'] when parsing bar()
['function', 'data', 'class', 'exception', 'method', 'classmethod', 'staticmethod', 'attribute', 'property', 'module', 'method']
```

### conf.py

```python
extensions = ['sphinx.ext.intersphinx']
intersphinx_mapping = {'python': ('https://docs.python.org/3',)}
``` 

### index.rst

```rst

.. py:function:: foo(x)
   
   The foo function.
   
   :param x: Some x.
   :type x: typing.Any

.. py:function:: bar(a)
   
   The bar function.
   
   :param a: Some a.
   :type a: typing.Any
```

### Environment Information

```text
Platform:              linux; (Linux-5.3.18-lp152.106-default-x86_64-with-glibc2.26)
Python version:        3.10.3 (main, Jan 31 2023, 10:47:25) [GCC 7.5.0])
Python implementation: CPython
Sphinx version:        6.2.0+/b04fe6f89
Docutils version:      0.18.1
Jinja2 version:        3.1.2
Pygments version:      2.15.0
```

### Sphinx extensions

```python
['sphinx.ext.intersphinx']
```

### Additional context

- #11282

### PR DESCRIPTION
**Pull request by picnixz**
Fix #11337 

### Discussion

I resisted making `domain.objtypes_for_role` return a `tuple` instead of a list to ensure that everything is read-only (it appears that occurrences of `objtypes` are only used as read-only, except in intersphinx) and expose a setter to update the corresponding object types given a specific role but I decided not to.

### Relates
- #11282

### CODE REVIEW

File: sphinx/ext/intersphinx.py [hunk 1]
```diff
@@ -358,16 +358,21 @@ def _resolve_reference_in_domain(env: BuildEnvironment,
...
+    if domain.name == 'py' and 'attribute' in objtypes and 'method' not in objtypes:
         # properties are stored as py:method since Sphinx 2.1
-        objtypes.append('method')
+        extra_objtypes.append('method')
 
     # the inventory contains domain:type as objtype
-    objtypes = [f"{domain.name}:{t}" for t in objtypes]
+    domain_name = domain.name
+    objtypes = [f"{domain_name}:{t}" for t in objtypes]
+    objtypes.extend([f'{domain_name}:{t}' for t in extra_objtypes])
```
Comment (AA-Turner) [inline sphinx/ext/intersphinx.py]:
```suggestion
    objtypes = [f"{domain_name}:{typ}" for typ in objtypes + extra_objtypes]
```

**Comment by AA-Turner** [comment]
I've updated to use dict-based set operations, to preserve order (I can't work out if order preservation is important to intersphinx---I don't think so, but I'm not confident).

Let me know thoughts on this approach!

A

File: sphinx/ext/intersphinx.py [hunk 2]
```diff
@@ -410,30 +412,30 @@ def _resolve_reference(env: BuildEnvironment, inv_name: str | None, inventory: I
...
+
+    domain_name = node.get('refdomain')
+    if not domain_name:
+        # only objects in domains are in the inventory
+        return None
+    if honor_disabled_refs \
+            and (domain_name + ":*") in env.config.intersphinx_disabled_reftypes:
+        return None
+    domain = env.get_domain(domain_name)
+    objtypes = {}.fromkeys(domain.objtypes_for_role(typ)).keys()
```
Comment (picnixz) [inline sphinx/ext/intersphinx.py]: We could defer the creation of the dictionary:

```python
_objtypes = domain.objtypes_for_role(typ)
if not _objtypes:
    return None
objtypes = {}.fromkeys(_objtypes).keys()
```

Comment (AA-Turner) [reply sphinx/ext/intersphinx.py]: This was included in https://github.com/sphinx-doc/sphinx/pull/11338/commits/6a9e617a49041b042d1666ba49a726f06aa46601, if I understand you correctly!

A

**Pull request merged**